Replicator/Worker.cc

```diff
@@ -313,10 +313,37 @@ namespace litecore { namespace repl {
             _parent = nullptr;
     }
 
-    void Worker::setMsgCollection(blip::MessageBuilder& msg, CollectionIndex i) {
+    // Either there is error, or return a valid collection index
+    std::pair<CollectionIndex, slice>
+    Worker::checkCollectionOfMsg(const blip::MessageIn& msg, CollectionIndex i) const {
```

Comment:
Personally I like a conservative function that does or returns one thing such as :

```
bool validateMessageCollectionIndex(CollectionIndex index, int refRequestIndex, C4Error* NULLABLE outError);
```

Then 
```
C4Error error;
CollectionIndex collIndex = getCollectionIndex(msg);
if (!validateMessageCollectionIndex(collIndex, -1, &error)) {
    ...
}
```

Reply:
I agree with this because a new smell is that the great majority of these function calls are ignoring the first returned argument from this function.  If it is being ignored more than used, something is wrong.

Reply:
getCollectionIndex and checkCollectionOfMsg are different. getCollectionIndex is at low level. It returns the index verbatim, except the handling of the nuance that the property is stored in message by long, but we are using it as unsigned. (typing here, I realized that I forgot to check the sign. I will change the assertion to
    Assert(0 <= l && l <= kNotCollectionIndex);
)
The value of kNotCollectionIndex will be interpreted at a higher level. For example, checkCollectionOfMsg() will never return kNotCollectionIndex unless there is an error. Another place it (kNotCollectionIndex) is interpreted is when the getCheckpoint message is received, as I try to tell whether it comes from 3.0 or 3.1 client.

Reply:
The reason I return error as slice rather than C4Error is because it will used in the context of request and response. For the former, we generate a local error as C4Error, but for that latter, we put the error in the response that we send back to the remote.

Reply:
The issue for me is the lack of use of the first argument.  More often it is simply "ignore" which should be the exception rather than the norm.  The code should not just work, it should also be easy to reason about.  If we, as reviewers, have this many questions then what will an outside person trying to understand the code think?

Reply:
Oh, just noticed it. I think you mean argument 'i' of CollectionIndex. I take it and will make the following changes in the next PR: changing it to two overloaded functions: one having 1 argument, and the other having 2 arguments. This better reflects the meaning of the functions.